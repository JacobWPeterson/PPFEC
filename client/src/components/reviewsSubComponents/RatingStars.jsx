/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import styled from 'styled-components';

const StarsContainer = styled.div`
  unicode-bidi: bidi-override;
  color: #c5c5c5;
  font-size: 22px;
  height: 22px;
  line-height: 22px;
  position: relative;
  padding: 0;
  text-shadow: 0px 1px 0 #a2a2a2;

  &.top {
    width: ${({ rating }) => (rating / 5) * 100}%;
    color: #FDD700;
    padding: 0;
    position: absolute;
    z-index: 1;
    display: block;
    top: 0;
    left: 0;
    overflow: hidden;
  }

  &.bottom {
    padding: 0;
    display: block;
    z-index: 0;
  }

`;

const InputStarWrapper = styled.div`
  fontSize: 25px;
  color: #FDD700;
  cursor: pointer;
`;

class RatingStars extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedStars: [1, 2, 3, 4, 5],
      rating: this.props.rating || 0,
      hovered: 0,
      selectedIcon: '★',
      deselectedIcon: '☆',
    };
    this.setRating = this.setRating.bind(this);
    this.hoverRating = this.hoverRating.bind(this);
  }

  setRating(inputRating) {
    this.setState({
      rating: inputRating,
    });

    if (this.props.getRating) {
      this.props.getRating(this.state.rating);
    }
  }

  hoverRating(rating) {
    this.setState({
      hovered: rating,
    });
  }

  render() {
    let staticStars;
    let inputStars;
    if (this.props.rating) {
      staticStars = (
        <StarsContainer>
          <StarsContainer className="top" rating={this.props.rating}>
            <span>★</span>
            <span>★</span>
            <span>★</span>
            <span>★</span>
            <span>★</span>
          </StarsContainer>
          <StarsContainer className="bottom">
            <span>★</span>
            <span>★</span>
            <span>★</span>
            <span>★</span>
            <span>★</span>
          </StarsContainer>
        </StarsContainer>
      );
    } else {
      inputStars = (
        <InputStarWrapper>
          {this.state.selectedStars.map((star) => {
            const hoverStars = this.state.hovered < star
              ? this.state.deselectedIcon : this.state.selectedIcon;
            return (
              <span
                key={star}
                onClick={() => { this.setRating(star); }}
                onMouseEnter={() => { this.hoverRating(star); }}
                onMouseLeave={() => { this.hoverRating(0); }}
              >
                {this.state.rating < star ? hoverStars : this.state.selectedIcon}
              </span>
            );
          })}
        </InputStarWrapper>
      );
    }

    return (
      <div>
        {staticStars}
        {inputStars}
      </div>
    );
  }
}

export default RatingStars;
