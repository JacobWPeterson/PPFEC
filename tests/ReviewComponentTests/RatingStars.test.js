/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import RatingStars from '../../client/src/components/reviewsSubComponents/RatingStars.jsx';

describe('RatingStars Component without passing props', () => {
  let RatingStarsComponent;
  beforeEach(() => {
    RatingStarsComponent = mount(<RatingStars debug />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should render correctly', () => {
    expect(RatingStarsComponent).toMatchSnapshot();
  });

  // it('Should have a onClick handler to set the rating', () => {
  //   const mockRating = 4;
  //   const spyOnClick = jest.spyOn(RatingStarsComponent.instance(), 'setRating');
  //   // RatingStarsComponent.simulate('Click');
  //   RatingStarsComponent.instance.setRating(mockRating);
  //   expect(spyOnClick).toHaveBeenCalled();
  //   expect(RatingStarsComponent.state('rating')).toBe(4);
  // });
});