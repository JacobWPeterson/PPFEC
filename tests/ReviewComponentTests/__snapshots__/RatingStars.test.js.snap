// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingStars Component without passing props Should render correctly 1`] = `
<RatingStars
  debug={true}
>
  <div>
    <RatingStars__InputStarWrapper>
      <div
        className="RatingStars__InputStarWrapper-sc-1djn8fm-1 kKUgkY"
      >
        <span
          key="1"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          ☆
        </span>
        <span
          key="2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          ☆
        </span>
        <span
          key="3"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          ☆
        </span>
        <span
          key="4"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          ☆
        </span>
        <span
          key="5"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          ☆
        </span>
      </div>
    </RatingStars__InputStarWrapper>
  </div>
</RatingStars>
`;
